force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,41 +227,67 @@ public with sharing class DMLOperations {
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
+            public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+                
+                List<Contact> contacts = new List<Contact>();
+                // Loop through each Contact
+                for(Contact contact : contactsToUpsert){
+                // Extract the Account Name from Contact's LastName
+                String accountName = contact.LastName;
 
-		// Extract the Account Name from Contact's LastName
+                // Get the account based on the name or create one if it doesn't exist
+                    Account account = upsertAccount(accountName);
 
-		// Get the account based on the name or create one if it doesn't exist
+                // Link the Contact to its Account using the Account Id
+                    contact.AccountId = account.Id;
 
-		// Link the Contact to its Account using the Account Id
+                    contacts.add(contact);
 
-		// Upsert the Contacts
-	}
+                // Upsert the Contacts
+            }
+            upsert contacts;
+        }
 
-    /**
+    /** Q11
      * Question 11
      * Insert a list of Leads and then deletes them.
      * @param leadNames The list of Lead names to be inserted.
      */
-    public static void insertAndDeleteLeads(List<String> leadNames) {
-        // Loop through each Lead name
-
-        // Insert the Leads
-
-        // Delete the Leads
-    }
-
-
-    /**
+        public static void insertAndDeleteLeads(List<String> leadNames) {
+            //list to hold the leads to be inserted
+            List<Lead> leadsToInsert = new List<Lead>();
+
+            // Loop through each Lead name
+            for (String leadName : leadNames){
+                // Create a Lead with the given name
+                Lead lead = new Lead(LastName = leadName, Company = 'Builders Vision');
+                
+                // Add the lead to the list
+                leadsToInsert.add(lead);
+            }
+
+                // Insert the Leads
+                insert leadsToInsert;
+
+                // Delete the Leads
+                delete leadsToInsert;
+        } 
+
+    /** Q12
      * Question 12
      * Dynamically create the number of cases and relate it to the accountId, then deletes them.
      * @param accountId The account for which the cases will be created and then deleted.
      * @param numOfCases The number of cases to create.
      */
-    public static void createAndDeleteCases(String accountId, Integer numOfCases) {
-        // Insert the cases
-
-        // Delete the cases
-    }
-}
\ No newline at end of file
+        public static void createAndDeleteCases(String accountId, Integer numOfCases) {
+            // Insert the cases
+            for (Integer i  = 0 ; i < numOfCases; i++){
+                Case caseToInsert = new Case(AccountId = accountId);
+                insert caseToInsert;
```

Comment:
🚩Here you're using a DML statement inside a loop. This goes against good practices and may result in exceeding Salesforce Governor limits. Instead, you need to bulkify this code by adding these cases to a List, and then using a single DML statement on the entire List outside the loop. (I think you know this already, but just making sure!)